[stronghold][bc-linter] Extract BC-linter workflow into a reusable action

izaitsevfb opened the pull request:
This PR extracts the BC-linter workflow from the PyTorch repository into a reusable GitHub Action in the `test-infra` repository. The rationale for this change is to simplify the management of the BC-linter workflow, promote code reusability, and make it easier to apply updates or improvements to the linter across multiple projects.

The new action can be found at `.github/actions/bc-lint/action.yml` in the `test-infra` repository.

# Usage

To use the BC-linter action in your workflow, add the following step to your GitHub Actions workflow file:

```yaml
- name: Run BC Lint Action
  uses: pytorch/test-infra/.github/actions/bc-lint@main
  with:
    repo: ${{ github.event.pull_request.head.repo.full_name }}
    base_sha: ${{ github.event.pull_request.base.sha }}
    head_sha: ${{ github.event.pull_request.head.sha }}
    suppression: ${{ contains(github.event.pull_request.labels.*.name, 'suppress-api-compatibility-check') }}
    docs_link: 'https://example.com/docs'
```

Replace `your_repo_path` with the path to the checked-out repository in your workflow, and `'https://example.com/docs'` with the actual link to the documentation.

# Inputs

The action takes the following inputs:

- `repo`: Github repo name to test
- `base_sha`: PR base SHA (required), use `${{ github.event.pull_request.base.sha }}`
- `head_sha`: PR head SHA (required), use `${{ github.event.pull_request.head.sha }}`
- `suppression`: Suppression flag (optional, default: ''). When 'true', the linter will issue warning and won't fail the job.
- `docs_link`: Link to the docs to display in case of failure (optional, default: '')

# Testing

Tested in a [separate repo](https://github.com/izaitsevfb/pr-head-test).

See [this run](https://github.com/izaitsevfb/pr-head-test/actions/runs/4886550214) for suppressed failure, and [this run](https://github.com/izaitsevfb/pr-head-test/actions/runs/4886181259/jobs/8721997250) for unsuppressed failure.

osalpekar commented on .github/actions/bc-lint/action.yml:
```diff
@@ -0,0 +1,65 @@
+name: 'BC Lint Action'
+description: 'A reusable action for running the BC Lint workflow. See tools/stronghold/README.md for more information.'
+inputs:
+  repo_path:
```
Why don't we just take the repo name as argument, and run the checkout of the specified repo in this reusable workflow (this is what we do in the Nova reusable workflows)? Makes the caller workflow cleaner as well since they don't have to checkout and pass around paths themselves.

osalpekar commented on .github/actions/bc-lint/action.yml:
```diff
@@ -0,0 +1,65 @@
+  suppression:
+    description: 'Suppression flag (true/false)'
+    required: false
+    default: ''
```
Should we default to `false`?

osalpekar reviewed (approved):
LGTM overall. Left 2 comments inline.

izaitsevfb commented:
Addressed review comment, tested the new version [in a separate repo](https://github.com/izaitsevfb/pr-head-test/actions/runs/4896705886/jobs/8743846641?pr=2).
